File: HMTargetActionListTests/HMWeakTargetActionTest.m
//
//  HMWeakTargetActionTest.m
//  HMTargetActionList
//
//  Created by Muronaka Hiroaki on 2015/09/04.
//  Copyright (c) 2015年 Muronaka Hiroaki. All rights reserved.
//

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import "HMWeakTargetAction.h"

@interface HMWeakTargetActionTest : XCTestCase

@property(nonatomic, strong) HMWeakTargetAction* forWeakTest;

@end

@implementation HMWeakTargetActionTest

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
    NSObject* obj = [NSObject new];
    self.forWeakTest = [[HMWeakTargetAction alloc] initWithTarget:obj action:@selector(setUp)];
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

/////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark -
#pragma mark test property

-(void)testProperties {
    HMWeakTargetAction* w1 = [[HMWeakTargetAction alloc] initWithTarget:self action:@selector(setUp)];
    
    XCTAssertEqual(w1.target, self);
    XCTAssertEqual(w1.action, @selector(setUp));
}

/////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark -
#pragma mark test equal or not equal

- (void)testEqual {
    HMWeakTargetAction* w1 = [[HMWeakTargetAction alloc] initWithTarget:self action:@selector(setUp)];
    HMWeakTargetAction* w2 = [[HMWeakTargetAction alloc] initWithTarget:self action:@selector(setUp)];
    
    XCTAssertEqual(w1, w1);
    XCTAssertEqualObjects(w1, w1);
    XCTAssertEqualObjects(w1, w2);
    
    // they aren't same object.
    XCTAssertNotEqual(w1, w2);
}

- (void)testNotEqualTarget {
    HMWeakTargetAction* w1 = [[HMWeakTargetAction alloc] initWithTarget:self action:@selector(setUp)];
    HMWeakTargetAction* w2 = [[HMWeakTargetAction alloc] initWithTarget:w1 action:@selector(setUp)];
    
    XCTAssertNotEqual(w1, w2);
    XCTAssertNotEqualObjects(w1, w2);
}

- (void)testNotEqualSelector {
    HMWeakTargetAction* w1 = [[HMWeakTargetAction alloc] initWithTarget:self action:@selector(setUp)];
    HMWeakTargetAction* w2 = [[HMWeakTargetAction alloc] initWithTarget:self action:@selector(tearDown)];
    
    XCTAssertNotEqual(w1, w2);
    XCTAssertNotEqualObjects(w1, w2);
}

/////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark -
#pragma mark test weak

-(void)testWeak {
    XCTAssertNil(self.forWeakTest.target);
}


- (void)testPerformanceExample {
    // This is an example of a performance test case.
    [self measureBlock:^{
        // Put the code you want to measure the time of here.
    }];
}

@end

File: Tests/HMTargetActionListTests/HMWeakTargetActionTest.swift
//
//  HMWeakTargetActionTest.swift
//
//  Created by Hiroaki Muronaka on 2021/10/08.

import XCTest
@testable import HMTargetActionList

final class HMWeakTargetActionTest: XCTestCase {
    var testObject: HMWeakTargetAction!
    
    override func setUp() {
        super.setUp()
        let obj = NSObject()
        self.testObject = HMWeakTargetAction(target: obj, action: #selector(HMWeakTargetActionTest.action))
    }
    
    //MARK: ///////////////////////////////////////////////////////////////////////////////////////////////
    //MARK: test properties
    
    func testProperties() {
        let w1 = HMWeakTargetAction(target: self, action: #selector(HMWeakTargetActionTest.action))
        
        XCTAssertEqual(w1.target as! HMWeakTargetActionTest, self)
        XCTAssertEqual(w1.action, #selector(HMWeakTargetActionTest.action))
    }
    
    //MARK: ///////////////////////////////////////////////////////////////////////////////////////////////
    //MARK: equals
    
    func testEquals() {
        let w1 = HMWeakTargetAction(target: self, action: #selector(HMWeakTargetActionTest.action))
        let w2 = HMWeakTargetAction(target: self, action: #selector(HMWeakTargetActionTest.action))
        
        XCTAssertEqual(w1, w2)
    }
    
    func testTargetIsNotEqual() {
        let w1 = HMWeakTargetAction(target: self, action: #selector(HMWeakTargetActionTest.action))
        let obj = NSObject()
        let w2 = HMWeakTargetAction(target: obj, action: #selector(HMWeakTargetActionTest.action))
        XCTAssertNotEqual(w1, w2)
    }
    
    func testSelectorIsNotEqual() {
        let w1 = HMWeakTargetAction(target: self, action: #selector(HMWeakTargetActionTest.action))
        let w2 = HMWeakTargetAction(target: self, action: #selector(HMWeakTargetActionTest.action2))
        XCTAssertNotEqual(w1, w2)
    }
    
    //MARK: weak test
    
    func testWeak1() {
        XCTAssertNil(self.testObject.target)
    }
    
    func testWeak2() {
        let w1 = HMWeakTargetAction(target: NSObject(), action: #selector(HMWeakTargetActionTest.action))
        XCTAssertNil(w1.target)
    }
    
    func testWeak3() {
        let obj = NSObject()
        let w1 = HMWeakTargetAction(target: obj, action: #selector(HMWeakTargetActionTest.action))
        XCTAssertEqual(w1.target as! NSObject, obj)
    }
}


extension HMWeakTargetActionTest {
    @objc func action() {
    }
    
    @objc func action2() {
    }
}
